### DISCUSSION TITLE
Model traversal

### DISCUSSION THREAD

**Discussion by scf37** [com-lihaoyi/scalatags#277] [General]
One of strong points of scalatags is ability to traverse, extend and transform assembled model before rendering. Unfortunately it is hacky right now - not everything is public and not everything is designed to be accessed by user code.

It could be nice to cleanup scalatags model to be proper AST (but still open for extension). Use cases:
- collect used resources and css classes before rendering
- declare custom tags and attributes to further control rendering. For example, adding selenium anchors only when tests are requesting the page
- transform model before rendering. I use that for CSS renaming.

**Comment by PerWiklander** [comment]
Other use cases:

 - Extracting dependencies for web components and adding them to `<head>`.
 - Checking that needed dependencies, like css classes, exists.

**Comment by lihaoyi** [comment]
This is definitely a possible use case, but not one that Scalatags supports right now. 

As written, Scalatags is very much designed to be "write only": you create a bunch of in-memory data structure, and stream it out to a `OutputStream` or a `dom.Element` (on Scala.js). Abstract is meant to be done at creation time (e.g. by passing in parameters) rather than post-creation transforamtions

It's definitely possible to insert an AST in there, which would perhaps not be that difficult given our existing support for custom backends. But it would take some work designing that AST to make it ergonomic and useful
